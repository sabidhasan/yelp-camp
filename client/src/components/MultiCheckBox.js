import React from 'react'
import PropTypes from 'prop-types'

class MultiCheckBox extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      // Select all items by default
      selectedIndices: props.items.map((v, i) => i),
      // this variable determines wherther to do "selectAll" or selectNone
      selectAll: false
    }

    this.updateValue = this.updateValue.bind(this)
    this.toggleSelect = this.toggleSelect.bind(this)
  }

  toggleSelect(e) {
    e.preventDefault();
    // if true, make false
    var newIndices = [];
    if (this.state.selectAll === true) {
      //Make new indices full
      for (let i = 0; i < this.props.items.length; i++) {
        newIndices.push(i)
      }
    }
    this.setState({
      selectAll: !this.state.selectAll,
      selectedIndices: newIndices
    });
    // Fire event listener
    if (this.props.onChange) {
      this.props.onChange(newIndices.map(v => this.props.items[v]))
    }
  }

  updateValue(event, newValue) {
    // Copy old state
    const oldState = this.state.selectedIndices.slice();
    const oldId = oldState.findIndex(val => val === newValue);
    if (oldId === -1) {
      // Add to state
      oldState.push(newValue)
    } else {
      // Remove from state
      oldState.splice(oldId, 1)
    }
    this.setState({selectedIndices: oldState});
    // Fire event listener
    this.props.onChange(oldState.map(v => this.props.items[v]))
  }

  render() {
    return (
    <React.Fragment>
      <p className='bold'>
        <span className='MultiCheckBox__title-text'>{this.props.defaultText} </span>
        <span className='MultiCheckBox__title-length'>
          ({this.state.selectedIndices.length}/{this.props.items.length} Selected)
        </span>
      </p>
      <button onClick={this.toggleSelect} className='btn btn--flat'>
        {`Select ${this.state.selectAll ? 'All' : 'None'}`}
      </button>
      <div className='MultiCheckBox__container'>
        {this.props.items
          .map((v, i) => {
            const checked = this.state.selectedIndices.includes(i) ? 'checked' : ''
            return (
              <label
                className='MultiCheckBox__label'
                key={i}
              >
                <input
                  className={`MultiCheckBox__checkbox ${checked ? 'MultiCheckBox__checkbox--checked' : ''}`}
                  type='checkbox'
                  value={v}
                  aria-describedby="MultiCheckBox__text"
                  aria-checked={checked}
                  checked={checked}
                  onChange={(event) => this.updateValue(event, i)} />
                <span className='MultiCheckBox__text' id='MultiCheckBox__text'>{v}</span>
              </label>
            )})
        }
      </div>
    </React.Fragment>
  )}
}

MultiCheckBox.propTypes = {
  items: PropTypes.array.isRequired,
  onChange: PropTypes.func.isRequired,
  defaultText: PropTypes.string
}

MultiCheckBox.defaultProps = {
  defaultText: ''
}

export default MultiCheckBox
