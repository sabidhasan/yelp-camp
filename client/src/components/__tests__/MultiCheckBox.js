import React from 'react'
import toJson from 'enzyme-to-json'
import { configure, shallow } from 'enzyme'
import Adapter from 'enzyme-adapter-react-16'
import MultiCheckBox from '../MultiCheckBox'

configure({ adapter: new Adapter() });

describe('<MultiCheckBox />', () => {
  let component;
  let props;
  const onChange = jest.fn();

  beforeEach(() => {
    props = {
      items: ['First', 'Second', 'Third'],
      defaultText: 'Test Text',
      onChange
    }
    component = shallow(<MultiCheckBox { ...props } />)
  })

  it('Component should be defined', () => {
    expect(component).toBeDefined();
  });

  it('', () => {

  })

  it('Default text should be displayed', () => {
    expect(component.find('.MultiCheckBox__title-text').text()).toBe('Test Text ')
  })

  it('Length of options shown correctly', () => {
    expect(component.find('.MultiCheckBox__title-length').text()).toBe('(3/3 Selected)')
  })

  it('Button initially says `Select None`', () => {
    expect(component.find('button').text()).toBe('Select None')
  })

  it('Number of checkboxes are correct for items given', () => {
    expect(component.find('.MultiCheckBox__label')).toHaveLength(3)
  })

  it('All items are initially selected with proper class name', () => {
    expect(component.find('.MultiCheckBox__checkbox--checked')).toHaveLength(3)
  })

  it('Callback function called when checkbox checked/unchecked', () => {
    component.find('.MultiCheckBox__checkbox--checked').first().simulate('change', {})
    expect(onChange).toHaveBeenCalledTimes(1);
  })

  it('Checkbox tickbox becomes unselected when unchecked', () => {
    component.find('.MultiCheckBox__checkbox--checked').first().simulate('change', {})
    expect(component.find('.MultiCheckBox__checkbox--checked')).toHaveLength(2);
  })

  it('Toggle button initially unselects all items', () => {
    component.find('button').simulate('click', {preventDefault: () => {}});
    expect(component.find('.MultiCheckBox__checkbox--checked')).toHaveLength(0);
  })

  it('When Toggle button is clicked, it changes text to `Select All`', () => {
    component.find('button').simulate('click', {preventDefault: () => {}});
    expect(component.find('button').first().text()).toBe('Select All');
  })

  it('Pressing toggle button twice reselects all items', () => {
    component.find('button').simulate('click', {preventDefault: () => {}});
    component.find('button').simulate('click', {preventDefault: () => {}});
    expect(component.find('.MultiCheckBox__checkbox--checked')).toHaveLength(3);
  })

  it('MultiCheckBox component snapshot matches what is expected', () => {
    expect(toJson(component)).toMatchSnapshot();
  })
})
