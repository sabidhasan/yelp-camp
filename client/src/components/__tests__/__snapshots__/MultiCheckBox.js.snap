// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MultiCheckBox /> MultiCheckBox component snapshot matches what is expected 1`] = `
<Fragment>
  <p
    className="bold"
  >
    <span
      className="MultiCheckBox__title-text"
    >
      Test Text
       
    </span>
    <span
      className="MultiCheckBox__title-length"
    >
      (
      3
      /
      3
       Selected)
    </span>
  </p>
  <button
    className="btn btn--flat"
    onClick={[Function]}
  >
    Select None
  </button>
  <div
    className="MultiCheckBox__container"
  >
    <label
      className="MultiCheckBox__label"
      key="0"
    >
      <input
        aria-checked="checked"
        aria-describedby="MultiCheckBox__text"
        checked="checked"
        className="MultiCheckBox__checkbox MultiCheckBox__checkbox--checked"
        onChange={[Function]}
        type="checkbox"
        value="First"
      />
      <span
        className="MultiCheckBox__text"
        id="MultiCheckBox__text"
      >
        First
      </span>
    </label>
    <label
      className="MultiCheckBox__label"
      key="1"
    >
      <input
        aria-checked="checked"
        aria-describedby="MultiCheckBox__text"
        checked="checked"
        className="MultiCheckBox__checkbox MultiCheckBox__checkbox--checked"
        onChange={[Function]}
        type="checkbox"
        value="Second"
      />
      <span
        className="MultiCheckBox__text"
        id="MultiCheckBox__text"
      >
        Second
      </span>
    </label>
    <label
      className="MultiCheckBox__label"
      key="2"
    >
      <input
        aria-checked="checked"
        aria-describedby="MultiCheckBox__text"
        checked="checked"
        className="MultiCheckBox__checkbox MultiCheckBox__checkbox--checked"
        onChange={[Function]}
        type="checkbox"
        value="Third"
      />
      <span
        className="MultiCheckBox__text"
        id="MultiCheckBox__text"
      >
        Third
      </span>
    </label>
  </div>
</Fragment>
`;
